analyses/unit test demo.Rmd

```diff
@@ -0,0 +1,230 @@
+---
+title: "Unit Test Demo"
+output: html_document
+date: "2023-09-14"
+---
+
+```{r setup, include=FALSE}
+library(tcplfit2)
+library(stringr)
+library(testthat)
+```
+
+The purpose of having unit tests is to ensure future feature development does not affect current analyses. How unit tests work is that it compares the results from the code (usually after some changes made) with previous known results to ensure the output is not changed when it's not supposed to. 
+
+Use `fitexp3` as an example, the current fit for exponential 3 with some simulated data is as below, and we want to make sure this does not change as we add more changes to the package.  
+
+```{r}
+data("signatures")
+conc=as.numeric(str_split(signatures[1,"conc"],"\\|")[[1]])
+resp=as.numeric(str_split(signatures[1,"resp"],"\\|")[[1]])
+# use this result as the "expected" result
+fitexp3(conc, resp)
+```
+
+To create a test for a function, we can use `test_that` function from `testthat` package. The test will run the code above and compare the result to what we expect to see (results from above). The following chunk is the test created for `fitexp3` and it checks for the fitted values for all model parameters. This code can be ran directly and the output would be either you pass or did not pass the test. 
+
+```{r}
+test_that("fitexp3 works", {
+  
+  # include the same codes that expect to generate the same results 
+  data("signatures")
+  conc=as.numeric(str_split(signatures[1,"conc"],"\\|")[[1]])
+  resp=as.numeric(str_split(signatures[1,"resp"],"\\|")[[1]])
+  
+  # compare the actual result with the expected 
+  expect_equal(fitexp3(conc, resp)$a, 2.786, tolerance = 1e-3)
+  expect_equal(fitexp3(conc, resp)$b, 269.522, tolerance = 1e-3)
+  expect_equal(fitexp3(conc, resp)$p, 0.6895, tolerance = 1e-3)
+})
+```
+
+We can also use `devtools::test()` to run all tests in the package at once. The result will show have many tests are ran and how many passed. For any failed tests, the result will show you the what is the actual output and what is the expected output.
+
+```{r}
+# make sure you are in the package directory, or specify the path to it
+devtools::test()
+```
+
+Let say I'm working on adding a new argument to allow users to choose normal as error distribution. The default for this argument would still be t-distribution as is right now, and adding this feature should not change the fitting result if normal error is not used. 
+
+For instance, I accidentally hard code the `fitexp3` to use normal for error distribution, it will return different fitted parameter values and cause the test to fail. 
+
+```{r}
```

Comment:
Please make this chunk scrollable, and in the text above I would mention which 'code section' the hard coding occurs (i.e. "Optimize the model").